swarm/pss/pss_new_test.go

```diff
@@ -0,0 +1,426 @@
+		params.MinProxBinSize = minProxBinSize
+		params.MaxBinSize = 3
+		params.MinBinSize = 1
+		params.MaxRetries = 1000
+		params.RetryExponent = 2
+		params.RetryInterval = 1000000
+		kademlias[id] = network.NewKademlia(id[:], params)
+		return kademlias[id]
+	}
+	return map[string]simulation.ServiceFunc{
+		"pss": func(ctx *adapters.ServiceContext, b *sync.Map) (node.Service, func(), error) {
+			// execadapter does not exec init()
+			initTest()
+
+			// create keys in whisper and set up the pss object
+			ctxlocal, cancel := context.WithTimeout(context.Background(), time.Second)
+			defer cancel()
+			keys, err := wapi.NewKeyPair(ctxlocal)
+			privkey, err := w.GetPrivateKey(keys)
+			pssp := NewPssParams().WithPrivateKey(privkey)
+			pssp.AllowRaw = allowRaw
+			pskad := kademlia(ctx.Config.ID)
+			ps, err := NewPss(pskad, pssp)
+			if err != nil {
+				return nil, nil, err
+			}
+
+			// register the handlers we've been passed
+			var deregisters []func()
+			for tpc, hndlrFunc := range handlerContextFuncs {
+				deregisters = append(deregisters, ps.Register(&tpc, hndlrFunc(ctx.Config)))
+			}
+
+			// if handshake mode is set, add the controller
+			// TODO: This should be hooked to the handshake test file
+			if useHandshake {
+				SetHandshakeController(ps, NewHandshakeParams())
+			}
+
+			// we expose some api calls for cheating
+			ps.addAPI(rpc.API{
+				Namespace: "psstest",
+				Version:   "0.3",
+				Service:   NewAPITest(ps),
+				Public:    false,
+			})
+
+			// return Pss and cleanups
+			return ps, func() {
+				// run the handler deregister functions in reverse order
+				for i := len(deregisters); i > 0; i-- {
+					deregisters[i-1]()
+				}
+			}, nil
+		},
+		"bzz": func(ctx *adapters.ServiceContext, b *sync.Map) (node.Service, func(), error) {
+			// normally translation of enode id to swarm address is concealed by the network package
+			// however, we need to keep track of it in the test driver aswell.
+			// if the translation in the network package changes, that can cause thiese tests to unpredictably fail
+			// therefore we keep a local copy of the translation here
+			addr := network.NewAddr(ctx.Config.Node())
+			addr.OAddr = nodeIDToAddr(ctx.Config.Node().ID())
+
+			hp := network.NewHiveParams()
+			hp.Discovery = false
+			config := &network.BzzConfig{
+				OverlayAddr:  addr.Over(),
+				UnderlayAddr: addr.Under(),
+				HiveParams:   hp,
+			}
+			return network.NewBzz(config, kademlia(ctx.Config.ID), stateStore, nil, nil), nil, nil
+		},
+	}
+}
+
+// makes sure we create the addresses the same way in driver and service setup
+func nodeIDToAddr(id enode.ID) []byte {
+	return id.Bytes()
+}
+
+// temporary function for polling a "stable" network
+// stability here means no conns or drops in network within a "serenity" duration
+// timeout is max time to wait for a "stable" network
+func serenityNowPlease(sim *simulation.Simulation, serenity time.Duration, timeout time.Duration) bool {
```

Comment:
I really dont get this :) why would you need this? just use snapshots and no discovery

Reply:
Even if we use snapshot, the connections still have to be made in the object. The health check can fire already before all these connections are made.

If we want to reliably measure that messages end up in their correct neighborhoods, we need to make sure that the network has finished forming. Otherwise the po can be checked against the wrong depth (for example when a node only has `minProxBinSize` peers, this depth 0 but actually defined as healthy now) and the end result will be wrong.

Reply:
By the way, I've noticed that stability formation in the network even with mere 16 peers takes quite a bit of time. Several seconds in fact. And even then there might be seemingly spurious connects afterwards. Intuitively it doesn't seem right to me. That would be a different _issue_ of course.

Reply:
Coming to think of it; I guess it was the same before, that networks could be healthy if there were just two _known_ peers.

Reply:
Actually, if there was a guarantee that when the snapshot is done loading, and _before_ the simulation is run, then all the resulting peers from that snapshot will populate the kademlia as _known_, then perhaps waiting for the healthy state will suffice. I would have to review that this is the case, though, and maybe propose in a separate PR an amendment to the snapshot loading if it is not.

Reply:
all snapshot based tests rely on this assumption and i think `WaitTillHealth` makes sure of that. @janos is that right?

Reply:
As we discussed in the orange-lounge, WaitTillHealth ensures the health to be true, but not that the state of snapshot is fully loaded. We are using WaitTillHealth to wait, but it may not be the correct thing to do.